server/reload.go

```diff
@@ -788,6 +788,80 @@ func (s *Server) Reload() error {
 	return nil
 }
 
+// ReloadOptions applies any supported options from the provided Option
+// type. This returns an error if an option which doesn't
+// support hot-swapping was changed.
+func (s *Server) ReloadOptions(newOpts *Options) error {
```

Comment:
This is a lot of repeat with Reload(), so what I would suggest is that this one could be as-is (assuming it does the same than Reload()) and change Reload() to first load the configuration file and then invoke this new ReloadOptions() function. I think that processing config file could be done outside of server lock, something like:
```
func (s *Server) Reload() error {
	s.mu.Lock()
	configFile := s.configFile
	s.mu.Unlock()
	if configFile == "" {
		return errors.New("can only reload config when a file is provided using -c or --config")
	}

	newOpts, err := ProcessConfigFile(configFile)
	if err != nil {
		return err
	}

	return s.ReloadOptions(newOpts)
}
```

Reply:
Perfect! I agree completely. Doing things this way would also help answer my question on test coverage (as current tests on Reload() would pass through). The only reason I didn't do it this way the first time was my hesitancy to move things outside of the mutex, since things are usually bounded by mutexes for a reason (even though I couldn't immediately see why). 

I will update and squash this PR soon.

Reply:
@kozlovic I've made the changes suggested.